lang/shared/src/main/scala/com/wavesplatform/lang/Context.scala

```diff
@@ -0,0 +1,54 @@
+package com.wavesplatform.lang
+
+import Context._
+import com.wavesplatform.lang.Terms.TYPE
+import monix.eval.Coeval
+
+case class Context(typeDefs: Map[String, CustomType], letDefs: Defs, functions: Map[String, CustomFunction])
+
+object Context {
+
+  type Defs = Map[String, (TYPE, Coeval[Any])]
+
+  val empty = Context(Map.empty, Map.empty, Map.empty)
+
+  case class CustomType(name: String, fields: List[(String, TYPE)])
+
+  sealed trait CustomFunction {
+    val name: String
+    val args: List[(String, TYPE)]
+    val resultType: TYPE
+    def eval(args: List[Any]) : Either[String,resultType.Underlying]
+    val types : (List[TYPE], TYPE)
```

Comment:
`types` looks strange, because we have `args` and `resultType`. How about [signature](https://developer.mozilla.org/en-US/docs/Glossary/Signature/Function)? Also it is clearly separated from "crypto signature", because we are in the context of `CustomFunction`

Reply:
yes that's better